File: cpp/tracker.cpp
#include "bacterium.h"
#include <iostream>
#include <vector>
#include <Windows.h>

using namespace cv;
using namespace std;


// Note, inFrame is a GRAY_SCALE picture!!!
tracker::tracker(Mat inFrame, int width, int radius) {
    this->inFrame = inFrame.clone();
    bact.width = width;
    initwidth = width;
    lost = false;
    if (!init(radius)) {
        isinited = false;
        cerr << "Cannot initialize! Please set point manually" << endl;
    }
    else
        isinited = true;
}


bool tracker::init(int radius) {
    Mat judge;
    int width = inFrame.cols, height = inFrame.rows;
    int *cnt_row = new int[height], *cnt_column = new int[width];
    double min, max;
    Point center(0, 0);

    cout << width << "\t" << height << endl;
    minMaxLoc(inFrame, &min, &max);

    judge = inFrame >= ((uchar)max - 2);

    judge = judge / 255;
    
    // First stage, get approx point
    for (int i = 0; i < height; i++)
        cnt_row[i] = 0;
    for (int i = 0; i < width; i++)
        cnt_column[i] = 0;

    for (int i = 0; i < height; i++) {
        uchar* ptr = judge.ptr<uchar>(i);
        for (int j = 0; j < width; j++) {
            if (ptr[j] == 1) {
                cnt_row[i]++;
                cnt_column[j]++;
            }
        }
    }

    for (int i = 0, max = 0; i < height; i++) {
        if (cnt_row[i] > max) {
            max = cnt_row[i];
            center.y = i;
        }
    }
    
    for (int j = 0, max = 0; j < width; j++) {
        if (cnt_column[j] > max) {
            max = cnt_column[j];
            center.x = j;
        }
    }

    cout << "First time search, " << center << endl;
    //Second
    int cnt = 0;
    double sumx = 0, sumy = 0;
    for (int i = 0; i < judge.rows; i++) {
        uchar* ptr_p = judge.ptr<uchar>(i);
        for (int j = 0; j < judge.cols; j++) {
            if ((i - center.y) * (i - center.y) + (j - center.x) * (j - center.x)
                < radius * radius && ptr_p[j] == 1) {
                cnt++;
                sumx += j;
                sumy += i;
            }
        }
    }

    int resx = tint(sumx / (double)cnt);
    int resy = tint(sumy / (double)cnt);

    if ((resx - center.x) * (resx - center.x) +
        (resy - center.y) * (resy - center.y) > radius * radius / 4) {
        return true;
    }
    else {
        bact.pos.x = resx;
        bact.pos.y = resy;
    }

    cout << "Find bacterium at " << bact.pos << endl;

    set_region();

    delete[] cnt_column;
    delete[] cnt_row;

    return true;
}


bool tracker::setnewbacterium(bacterium bact) {
    this->bact = bact;
    set_region();
    return true;
}


bool tracker::update(Mat inframe) {
    double min, max;
    double min_global, max_global;
    double global_contrast;
    const int scale = 20;
    
    Rect focus_region(bact.rangel, bact.ranger);
    Mat focus_window = inframe(focus_region);

    // Get local region properties
    minMaxLoc(focus_window, &min, &max);

    double sumx = 0, sumy = 0;
    int cnt = 0;
    for (int i = 0; i < focus_window.rows; i++) {
        uchar* ptr = focus_window.ptr<uchar>(i);
        for (int j = 0; j < focus_window.cols; j++) {
            if (ptr[j] >= (int)max - scale) {
                sumx += j;
                sumy += i;
                cnt++;
            }
        }
    }

    bact.pos.x = tint(sumx / cnt) + bact.rangel.x;
    bact.pos.y = tint(sumy / cnt) + bact.rangel.y;

    {
        uchar* ptr = focus_window.ptr<uchar>(tint(sumy / cnt));
        int radius = bact.width, low = 0, high = 0;
        bool islowset = false;

        for (int j = 0; j < focus_window.cols; j++) {
            if (ptr[j] >= (int)max - scale) {
                if (!islowset) {
                    low = j;
                    islowset = true;
                }
                else
                    high = j;
            }
        }
        radius = high - low;
        if (radius > initwidth - 5)
            bact.width = radius + 5;
        else
            bact.width = initwidth;
    }

    if (bact.pos.x < bact.width || bact.pos.x > inframe.cols - bact.width ||
        bact.pos.y < bact.width || bact.pos.y > inframe.rows - bact.width) {
        lost = true;
        return false;
    }

    set_region();

    contrast = get_contrast(focus_window);

    // Get global properties
    // global_contrast = get_contrast(inframe);
    minMaxLoc(inframe, &min_global, &max_global);

    if (contrast <= 150 || max_global > 1.2 * max) {
        out_focus = true;
        return false;
    }
    else {
        out_focus = false;
        return true;
    }
}


void tracker::set_region() {

    // Set up range
    if (bact.pos.x - bact.width < 0)
        bact.rangel.x = 0;
    else
        bact.rangel.x = bact.pos.x - bact.width;

    if (bact.pos.y - bact.width < 0)
        bact.rangel.y = 0;
    else
        bact.rangel.y = bact.pos.y - bact.width;

    if (bact.pos.x + bact.width >= inFrame.cols)
        bact.ranger.x = inFrame.cols - 1;
    else
        bact.ranger.x = bact.pos.x + bact.width;

    if (bact.pos.y + bact.width >= inFrame.rows)
        bact.ranger.y = inFrame.rows - 1;
    else
        bact.ranger.y = bact.pos.y + bact.width;

}


double get_contrast(cv::Mat frame) {
    int width = frame.cols - 2;
    int height = frame.rows - 2;
    Mat inframe = frame.clone();

    Rect center_r(1, 1, width, height);
    Rect up_r(1, 0, width, height);
    Rect down_r(1, 2, width, height);
    Rect left_r(0, 1, width, height);
    Rect right_r(2, 1, width, height);

    Mat up;
    Mat down;
    Mat left;
    Mat right;
    Mat center;

    inframe(up_r).clone().convertTo(up, CV_32S);
    inframe(down_r).clone().convertTo(down, CV_32S);
    inframe(left_r).clone().convertTo(left, CV_32S);
    inframe(right_r).clone().convertTo(right, CV_32S);
    inframe(center_r).clone().convertTo(center, CV_32S);

    multiply((up - center), (up - center), up);
    multiply((down - center), (down - center), down);
    multiply((left - center), (left - center), left);
    multiply((right - center), (right - center), right);

    return (double)(sum(up + down + left + right)[0]) / (width * height * 1.0);
}

void onMouse(int event, int x, int y, int flags, void* para) {
    tracker* track = (tracker*)para;

    if (event == EVENT_LBUTTONDOWN) {
        Sleep(20);
        bacterium bact;
        bact.pos.x = x;
        bact.pos.y = y;
        bact.width = track->initwidth;

        track->setnewbacterium(bact);
        track->lost = false;
    }
}


int tint(double x) {
    if (x - (int)x >= 0.5)
        return (int)x + 1;
    else
        return (int)x;
}

File: python/tracker.py
import numpy as np
from time import sleep, clock
import cv2
import win32api
import socket
from bacterium import bacterium
import pyHook, pythoncom
from subprocess import call
from PIL import ImageGrab


def onMouse(event, x, y, flag, param):
    if isinstance(param, tracker) == False:
        return
    if event == cv2.EVENT_LBUTTONDOWN:
        param.changePos(2 * x, 2 * y)


def getContrast(inframe):
    a = len(np.shape(inframe))
    if len(np.shape(inframe)) != 2:
        return

    width, height = np.shape(inframe)
    width -= 2
    height -= 2

    center = inframe[1 : height, 1 : width]
    left = inframe[1 : height, 0 : width - 1]
    right = inframe[1 : height, 2 : width + 1]
    up = inframe[0 : height - 1, 1 : width]
    down = inframe[2 : height + 1, 1 : width]

    res = abs(center - left)  + abs(center - right) ** 2 + abs(center - up) ** 2\
          + abs(center - down) ** 2

    return res.mean()


class Region:
    def __init__(self):
        self.start = []
        self.end = []


class tracker():
    def __init__(self, bact, max_x, max_y):
        self.bact = bact
        if isinstance(bact, bacterium) == False:
            print("Error")
            return
        self.isinited = True
        self.outofFocus = False
        self.isLost = True
        self.xrange = [0, 0]
        self.yrange = [0, 0]
        self.max_x = max_x
        self.max_y = max_y
        self.ischangeing = False
        self.contrast_prev = 0
        self.dire_up = False
        self.isgreat = True

    def changePos(self, x, y):
        self.bact.x = x
        self.bact.y = y
        self.isLost = False
        self.outofFocus = False
        self.ischangeing = True
        self.setRange()
        self.ischangeing = False

    def update(self, frame):
        if self.isinited == False or self.isLost == True or self.ischangeing == True:
            return 0

        ylow = self.yrange[0]
        yhigh = self.yrange[1]
        xlow = self.xrange[0]
        xhigh = self.xrange[1]

        region = frame[ylow : yhigh, xlow: xhigh]

        # region = region.astype(np.double) ** 2
        # region = region - region.mean()
        # region = 255 * (region / region.max())
        # region = region.astype(np.uint8)

        res = np.where(region >= region.max() - 3)
        resy = res[0]
        resx = res[1]
        posx = resx.mean()
        posy = resy.mean()

        self.bact.x = self.xrange[0] + posx
        self.bact.y = self.yrange[0] + posy
        self.setRange()

        # print(self.xrange, self.yrange, self.bact.x, self.bact.y)

        contra = getContrast(region)
        if  contra< 30:
            self.outofFocus = True
        else:
            self.outofFocus = False
        return contra

    def auto_focus(self, contrast):
        if self.outofFocus == False:
            return

        if contrast > self.contrast_prev:
            self.isgreat = True
        else:
            self.isgreat = False

        if self.isgreat == True:
            if self.dire_up == True:
                self.movez(10)
            else:
                self.movez(-10)
        else:
            self.dire_up = not self.dire_up
            if self.dire_up == True:
                self.movez(10)
            else:
                self.movez(-10)


    def movez(self, step):
        call(["Stagez.exe", str(step)])

    def setRange(self):
        x = int(self.bact.x)
        y = int(self.bact.y)

        if x - self.bact.width > 0:
            self.xrange[0] = x - self.bact.width;
        else:
            self.isLost = True
            return

        if x + self.bact.width < self.max_x:
            self.xrange[1] = x + self.bact.width
        else:
            self.isLost = True
            return

        if y - self.bact.width > 0:
            self.yrange[0] = y - self.bact.width
        else:
            self.isLost = True
            return

        if y + self.bact.width < self.max_y:
            self.yrange[1] = y + self.bact.width
        else:
            self.isLost = True
            return


def LeftDown(event):
    print("here")
    global cnt
    global region
    if event.MessageName == 'mouse left down':
        if cnt == 0:
            region.start = event.Position
        elif cnt == 1:
            region.end = event.Position
        cnt += 1
    return True



if __name__ == '__main__':
    # font = cv2.FONT_HERSHEY_SIMPLEX
    # hm = pyHook.HookManager()
    # hm.MouseLeftDown = LeftDown
    # cnt = 0
    # region = Region()
    # hm.HookMouse()
    # while(cnt < 2):
    #     pythoncom.PumpWaitingMessages()

    # hm.UnhookMouse()
    # grab_r = (region.start[0], region.start[1], region.end[0], region.end[1])
    cap = cv2.VideoCapture(r"D:\Users\Neil Zhao\Documents\Tencent Files\1055306580\FileRecv\1021\great.avi")
    ret, frame = cap.read()
    y, x, channel = np.shape(frame)
    track = tracker(bacterium(30), grab_r[3]-grab_r[1], grab_r[2] - grab_r[0])
    # track = tracker(bacterium(30), x, y)
    cv2.namedWindow('test')
    cv2.setMouseCallback('test', onMouse, track)
    contrast = 0

    while(1):
        # ret, frame = cap.read()
        # if frame is None:
        #     break
        frame = ImageGrab.grab(bbox=(grab_r))
        frame = np.array(frame)
        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

        # gray = gray.astype(np.double) ** 2
        # gray = gray - gray.mean()
        # gray = 255 * (gray / gray.max())
        # gray = gray.astype(np.uint8)
        #
        # frame = frame.astype(np.double) ** 2
        # frame = frame - frame.mean()
        # frame = 255 * (frame / frame.max())
        # frame = frame.astype(np.uint8)
        # kernel = np.ones((track.max_y, track.max_y), np.float32) / (track.max_y * track.max_y)
        # gray = cv2.GaussianBlur(gray,(5, 5),0)
        # frame = cv2.GaussianBlur(frame, (5, 5), 0)

        contrast = track.update(gray)

        if track.outofFocus == True:
            cv2.putText(frame, "Out of focus", (10, 250), font, 1.2, (0, 255, 255), 2, cv2.LINE_AA)
            cv2.rectangle(frame, (track.xrange[0], track.yrange[0]), (track.xrange[1], track.yrange[1]), (255, 0, 0), 3)
            # track.auto_focus(contrast)
        elif track.isLost == True:
            cv2.putText(frame, "Bacterium lost", (10, 200), font, 1.2, (0, 0, 255), 2, cv2.LINE_AA)
        else:
            cv2.putText(frame, "2D fixed", (10, 200), font, 1.2, (0, 255, 0), 2, cv2.LINE_AA)
            cv2.rectangle(frame, (track.xrange[0], track.yrange[0]), (track.xrange[1], track.yrange[1]), (255, 0, 0), 3)

        cv2.putText(frame, "x=%.2f, y=%.2f, contrast=%.2f" % (track.bact.x, track.bact.y, contrast), (10, 50), font, 1.2,
                        (255, 255, 255), 2, cv2.LINE_AA)
        frame = cv2.resize(frame, None, fx=0.5, fy=0.5, interpolation = cv2.INTER_CUBIC)
        cv2.imshow('test', frame)
        if cv2.waitKey(30) & 0xFF == 27:
            break

    cv2.destroyAllWindows()